### DISCUSSION TITLE
Underlying simulation

### DISCUSSION THREAD

**Discussion by chaseltb** [IBM/aihwkit#687] [General] [answered]
## Description and motivation

To implement analog hardware, we need more information about how aihwkit simulates memristors and the pulses. Can anyone provide some insight into how the underlying crossbar/analog devices are represented? It would be incredibly useful to have the pulses as well as location of each memristor or components of the rpu so it can be implemented in hardware.

## Proposed solution

Adding information to rpu configuration and training. This could be in the form of a NumPy array or a few variables, and could have a boolean flag passed to the function to print them or just return them from a function.

## Alternatives and other information

Enabling the rpu configuration to be mapped to a data structure that can be accessed.

**Comment by anu-pub** [comment, marked as answer]
Thanks for the comment. Regarding how aihwkit simulates memristive devices and pulses, we already have documentation on that here for training https://aihwkit.readthedocs.io/en/latest/analog_update.html, and here for inference https://aihwkit.readthedocs.io/en/latest/pcm_inference.html. Also our APL tutorial paper has additional information https://pubs.aip.org/aip/aml/article/1/4/041102/2923573. 

For generating information about how layers are mapped, the AnalogInfo utility can be used: https://aihwkit.readthedocs.io/en/latest/api/aihwkit.utils.analog_info.html#aihwkit.utils.analog_info.AnalogInfo. The `create_layer_summary()` function creates a list of elements that correspond to each layer of the model, and provides various information including mapping. 

Regarding applied pulses during training, this is really hardware-specific, so it would not make sense that aihwkit provides waveforms or something similar as it would likely not work on the user's specific hardware. 

Please let us know if there is something not included in the above that you would like to have additional information on.
